Plugins/BuildServerIntegration/AzureDevOpsIntegration/AzureDevOpsAdapter.cs

```diff
@@ -144,97 +145,132 @@ private async Task ObserveBuildsAsync(DateTime? sinceDate, bool running, IObserv
+                        BuildInfo? buildInfo = CreateBuildInfo(buildToDisplay);
+                        observer.OnNext(buildInfo);
+                        _buildsCache.FinishedBuilds.Add(buildInfo);
+                        if (buildToDisplay.FinishTime.HasValue && buildToDisplay.FinishTime.Value >= _buildsCache.LastCall)
                         {
-                            ProjectOnErrorKey = CacheKey;
-                            MessageBox.Show(errorMessage, _buildIntegrationErrorCaption.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
+                            _buildsCache.LastCall = buildToDisplay.FinishTime.Value.AddSeconds(1);
                         }
-
-                        return;
                     }
                 }
+            }
+            catch (OperationCanceledException)
+            {
+                // Do nothing, the observer is already stopped
+            }
+            catch (Exception ex)
+            {
+                observer.OnError(ex);
+            }
+        }
 
-                void ManageError(string errorMessage)
-                {
-                    _apiClient = null;
-                    observer.OnNext(new BuildInfo { CommitHashList = new[] { ObjectId.WorkTreeId }, Description = errorMessage, Status = BuildInfo.BuildStatus.Failure });
-                    observer.OnCompleted();
-                }
+        private async Task<bool> EnsureBuildDefinitionsIsInitializedAsync(IObserver<BuildInfo> observer)
+        {
+            if (_buildDefinitions is not null)
+            {
+                return true;
+            }
+
+            try
+            {
+                Validates.NotNull(_buildDefinitionsTask);
+
+                _buildDefinitions = await _buildDefinitionsTask.JoinAsync();
 
                 if (_buildDefinitions is null)
                 {
                     observer.OnCompleted();
-                    return;
+                    return false;
                 }
 
-                var builds = running ?
-                    FilterRunningBuilds(await _apiClient.QueryRunningBuildsAsync(_buildDefinitions)) :
-                    await _apiClient.QueryFinishedBuildsAsync(_buildDefinitions, sinceDate);
+                _buildsCache = new BuildsCache { Id = CacheKey, BuildDefinitions = _buildDefinitions };
+                return true;
+            }
+            catch (UnauthorizedAccessException)
+            {
+                string errorMessage = _badTokenErrorMessage.Text;
+                ManageError(errorMessage);
 
-                foreach (var build in builds)
+                if (_projectOnErrorKey is null || _projectOnErrorKey != CacheKey)
                 {
-                    observer.OnNext(CreateBuildInfo(build));
+                    _projectOnErrorKey = CacheKey;
+
+                    TaskDialogButton btnOpenSettings = new("Open settings");
+                    TaskDialogButton btnIgnore = new("Ignore");
```

Comment:
Do we need to loc these?

Reply:
Done even if the translation is not supported for plugins...

Reply:
If I understand it correctly, the plugins that are bundled with the app are generally translated. Are we talking about the same thing?